### ISSUE TITLE
V0.2.4以上版本如何加载ptunig-v2的微调模型？

### ISSUE THREAD

**Issue by wvivi2023** [chatchat-space/Langchain-Chatchat#1671]
**问题描述 / Problem Description**
现在我仅仅用chatglm2 做ptuning v2 微调，现在想通过langchain-chatchat加载ptuning v2 微调出来的模型，不了解如何加载

**复现问题的步骤 / Steps to Reproduce**
1. [在](https://github.com/THUDM/ChatGLM-6B/tree/main/ptuning) 进行chatglm2 做ptuning v2 微
2. 微调出来的模型，不知道如何通过本项目加载

**预期的结果 / Expected Result**
方便加载ptuning v2 微调 模型

**实际结果 / Actual Result**
不知如何加载ptuning v2 微调 模型

**环境信息 / Environment Information**
- langchain-ChatGLM 版本/commit 号：[v0.2.4](https://github.com/chatchat-space/Langchain-Chatchat/releases/tag/v0.2.4）
- 是否使用 Docker 部署（是/否）：否
- 使用的模型（ChatGLM2-6B / Qwen-7B 等）： ChatGLM2-6B
- 使用的 Embedding 模型（moka-ai/m3e-base 等）：moka-ai/m3e-base 
- 使用的向量库类型 (faiss / milvus / pg_vector 等)：
- 操作系统及版本 / Operating system and version:linux
- Python 版本 / Python version:3.10.0
- 其他相关环境信息 / Other relevant environment information:

**附加信息 / Additional Information**
添加与问题相关的任何其他信息 / Add any other information related to the issue.

**Comment by huangyu996** [comment]
调用模型部分增加如下代码，微调输出的check point路径替换成你自己的
![image](https://github.com/chatchat-space/Langchain-Chatchat/assets/147224487/28e3c796-f629-4d24-b8df-e64a8dfdebd3)


tokenizer = AutoTokenizer.from_pretrained("/home/ai/models/chatglm2-6b", trust_remote_code=True)
config = AutoConfig.from_pretrained("/home/ai/models/chatglm2-6b", trust_remote_code=True, pre_seq_len=128)
model = AutoModel.from_pretrained("/home/ai/models/chatglm2-6b", config=config, trust_remote_code=True).cuda()
prefix_state_dict = torch.load(os.path.join("/data/out/self/checkpoint-100", "pytorch_model.bin"))
new_prefix_state_dict = {}
for k, v in prefix_state_dict.items():
if k.startswith("transformer.prefix_encoder."):
new_prefix_state_dict[k[len("transformer.prefix_encoder."):]] = v

**Comment by shujiepan** [comment]
请问修改的是哪个代码文件

**Comment by liuxjerry** [comment]
> 请问修改的是哪个代码文件

+1

**Comment by zRzRzRzRzRzRzR** [comment]
查看wiki

**Comment by XiaoXiaoJiangYun** [comment]
wiki里看不懂要改哪里才能加载ptuning v2微调后的模型？

**Comment by yuanzui0728** [comment]
同问

**Comment by shuzhiweb** [comment]
> 请问修改的是哪个代码文件

+1

**Comment by h-o-u-w-e-i** [comment]
我想作者是说去看https://github.com/chatchat-space/Langchain-Chatchat/wiki/%E6%9C%80%E4%BD%B3%E5%AE%9E%E8%B7%B5，因为这个项目本质上是通过fastchat加载本地模型，自己并没有加载逻辑，所以你要让fastchat兼容ptuning v2，目前似乎只能自己动手

**Comment by natureLanguageQing** [comment]
我找到了一个办法。在fastchat的路径下fastchat\model\model_adapter.py增加以下代码。`python
    def load_ptuning_model(self, model_path: str,ptuning_checkpoint :str, from_pretrained_kwargs: dict):
        revision = from_pretrained_kwargs.get("revision", "main")
        try:
            tokenizer = AutoTokenizer.from_pretrained(
                model_path,
                use_fast=self.use_fast_tokenizer,
                revision=revision,
                trust_remote_code=True,
            )
        except TypeError:
            tokenizer = AutoTokenizer.from_pretrained(
                model_path, use_fast=False, revision=revision, trust_remote_code=True
            )
        try:
            model = AutoModelForCausalLM.from_pretrained(
                model_path,
                low_cpu_mem_usage=True,
                trust_remote_code=True,
                **from_pretrained_kwargs,
            )
        except NameError:
            model = AutoModel.from_pretrained(
                model_path,
                low_cpu_mem_usage=True,
                trust_remote_code=True,
                **from_pretrained_kwargs,
            )
        prefix_state_dict = torch.load(os.path.join(ptuning_checkpoint, "pytorch_model.bin"))
        new_prefix_state_dict = {}
        for k, v in prefix_state_dict.items():
            if k.startswith("transformer.prefix_encoder."):
                new_prefix_state_dict[k[len("transformer.prefix_encoder."):]] = v
        model.transformer.prefix_encoder.load_state_dict(new_prefix_state_dict)
        model.transformer.prefix_encoder.float()

        return model, tokenizer
`

然后在fastchat\serve\model_worker.py的ModelWorker中加入参数 ptuning_checkpoint。
class ModelWorker(BaseModelWorker):
    def __init__(
        self,
        controller_addr: str,
        worker_addr: str,
        worker_id: str,
        model_path: str,
        model_names: List[str],
        limit_worker_concurrency: int,
        no_register: bool,
        device: str,
        num_gpus: int,
        max_gpu_memory: str,
        revision: str = None,
        ptuning_checkpoint: str =None,
        dtype: Optional[torch.dtype] = None,
        load_8bit: bool = False,
        cpu_offloading: bool = False,
        gptq_config: Optional[GptqConfig] = None,
        awq_config: Optional[AWQConfig] = None,
        exllama_config: Optional[ExllamaConfig] = None,
        xft_config: Optional[XftConfig] = None,
        stream_interval: int = 2,
        conv_template: Optional[str] = None,
        embed_in_truncate: bool = False,
        seed: Optional[int] = None,
        debug: bool = False,
接下来 加入以下的代码用来加载ptuning模型。

**Comment by natureLanguageQing** [comment]
controller_addr: str - 这是工作者将与之通信的控制器的地址。它是一个字符串，很可能指定了控制器在网络上的位置。
worker_addr: str - 这是工作者的地址，用于在分布式系统中标识和定位工作者。
worker_id: str - 工作者的唯一标识符，用于在系统中区分不同的工作者。
model_path: str - 模型文件的路径，工作者将加载和运行这个模型。
model_names: List[str] - 模型的名称列表，可能用于识别或选择不同的模型。
limit_worker_concurrency: int - 限制工作者并发执行的数量，可能用于控制系统资源的使用。
no_register: bool - 是否不注册工作者到控制器，这可能是控制工作者是否需要在系统中注册的一个标志。
device: str - 指定模型将在哪种设备上运行，如CPU或GPU。
num_gpus: int - 指定工作者拥有的GPU数量。
max_gpu_memory: str - 指定工作者可以使用的最大GPU内存量。
revision: str = None - 可能是指模型的版本或修订号，如果未指定则使用默认值None。
ptuning_checkpoint: str =None - 可能是指用于模型参数微调的检查点文件的路径，如果未指定则使用默认值None。
dtype: Optional[torch.dtype] = None - 指定模型的数据类型，如果未指定则使用默认值None。
load_8bit: bool = False - 是否以8位精度加载模型，这可能是为了减少内存使用和提高推理速度。
cpu_offloading: bool = False - 是否将计算任务从GPU卸载到CPU，这可能是为了在GPU负载过高时平衡系统负载。
gptq_config: Optional[GptqConfig] = None - 可能是指模型灰度发布配置，如果未指定则使用默认值None。
awq_config: Optional[AWQConfig] = None - 可能是指模型自动调优配置，如果未指定则使用默认值None。
exllama_config: Optional[ExllamaConfig] = None - 可能是指模型扩展配置，如果未指定则使用默认值None。
xft_config: Optional[XftConfig] = None - 可能是指模型交叉特征转换配置，如果未指定则使用默认值None。
stream_interval: int = 2 - 指定某种流式处理或数据流的间隔时间，单位可能是秒或其他时间单位。
conv_template: Optional[str] = None - 可能是指模型卷积模板的路径或定义，如果未指定则使用默认值None。
embed_in_truncate: bool = False - 是否在嵌入截断时嵌入某些信息，这可能是为了在模型处理输入时保留某些关键信息。
seed: Optional[int] =None - 指定随机种子，用于确保可重复的随机序列，这对于调试和测试非常有用。
debug: bool = False - 是否处于调试模式，这可能会影响模型的行为和系统日志的记录。
